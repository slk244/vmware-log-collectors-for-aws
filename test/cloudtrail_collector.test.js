/*
Copyright 2018 VMware, Inc.
SPDX-License-Identifier: MIT
*/

const { lintTestEnv, sendLogsAndVerify } = require('../test/helper.test');
const { createSample1 } = require('./sample_files/cloudtrail_testdata');

const {
  CloudTrailHttpCollector,
  CloudTrailKafkaCollector,
} = require('../src/index');

describe('CloudTrailKafkaCollector', () => {
  const collector = new CloudTrailKafkaCollector(lintTestEnv);

  it('processLogsJson', () => {
    const logsJson = createSample1();
    collector.processLogsJson(logsJson);
    expect(logsJson.Records).toBe(undefined);
    expect(logsJson).toMatchSnapshot();
  });
});

describe('CloudTrailHttpCollector', () => {
  const collector = new CloudTrailHttpCollector(lintTestEnv);

  describe('processLogsJson', () => {
    it('should process logs JSON', () => {
      const logsJson = createSample1();
      collector.processLogsJson(logsJson);
      expect(logsJson).toMatchSnapshot();
    });
  });

  describe('sendLogs', () => {
    it('should send request to the HTTP stream', (done) => {
      const logsJson = {
        Records: [
          {
            id: 'id1',
            field1: 'value1',
          },
        ],
      };

      const expectedReqHeaders = {
        reqheaders: {
          authorization: 'Bearer mocktoken',
          structure: 'simple',
          'content-type': 'application/json',
        },
      };

      const expectedReqBody = [{
        id: 'id1',
        field1: 'value1',
        text: '{"id":"id1","field1":"value1"}',
        ingest_timestamp: 1538769915817,
        log_type: 'aws_cloud_trail',
      }];

      sendLogsAndVerify(done, collector, logsJson, expectedReqBody, expectedReqHeaders);
    }, 10000);
  });
});
