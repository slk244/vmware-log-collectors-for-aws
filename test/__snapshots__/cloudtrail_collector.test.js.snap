// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudTrailHttpCollector processLogsJson should process logs JSON 1`] = `
Object {
  "Records": Array [
    Object {
      "awsRegion": "ca-central-1",
      "eventID": "c9931287-acd4-478f-aa24-f77c8d297f71",
      "eventName": "DescribeSubnets",
      "eventSource": "ec2.amazonaws.com",
      "eventTime": "2018-06-14T17:17:56Z",
      "eventType": "AwsApiCall",
      "eventVersion": "1.05",
      "ingest_timestamp": 1538769915817,
      "log_type": "aws_cloud_trail",
      "recipientAccountId": "unittest9",
      "requestID": "03c4406b-42dc-4172-8beb-d76e15b9d69f",
      "requestParameters_filterSet_items": Array [
        Object {
          "name": "vpc-id",
          "valueSet": Object {
            "items": Array [
              Object {
                "value": "vpc-04e3326d",
              },
            ],
          },
        },
      ],
      "responseElements": null,
      "sourceIPAddress": "13.56.178.189",
      "text": "{\\"eventVersion\\":\\"1.05\\",\\"userIdentity\\":{\\"type\\":\\"IAMUser\\",\\"principalId\\":\\"AIDAIRUNBBKT2UKVMBDMM\\",\\"arn\\":\\"arn:aws:iam::792677813823:user/flands\\",\\"accountId\\":\\"792677813823\\",\\"accessKeyId\\":\\"AKIAJDNI5LT67QRAUMIQ\\",\\"userName\\":\\"flands\\"},\\"eventTime\\":\\"2018-06-14T17:17:56Z\\",\\"eventSource\\":\\"ec2.amazonaws.com\\",\\"eventName\\":\\"DescribeSubnets\\",\\"awsRegion\\":\\"ca-central-1\\",\\"sourceIPAddress\\":\\"13.56.178.189\\",\\"userAgent\\":\\"aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8\\",\\"requestParameters\\":{\\"subnetSet\\":{},\\"filterSet\\":{\\"items\\":[{\\"name\\":\\"vpc-id\\",\\"valueSet\\":{\\"items\\":[{\\"value\\":\\"vpc-04e3326d\\"}]}}]}},\\"responseElements\\":null,\\"requestID\\":\\"03c4406b-42dc-4172-8beb-d76e15b9d69f\\",\\"eventID\\":\\"c9931287-acd4-478f-aa24-f77c8d297f71\\",\\"eventType\\":\\"AwsApiCall\\",\\"recipientAccountId\\":\\"unittest9\\"}",
      "userAgent": "aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8",
      "userIdentity_accessKeyId": "AKIAJDNI5LT67QRAUMIQ",
      "userIdentity_accountId": "792677813823",
      "userIdentity_arn": "arn:aws:iam::792677813823:user/flands",
      "userIdentity_principalId": "AIDAIRUNBBKT2UKVMBDMM",
      "userIdentity_type": "IAMUser",
      "userIdentity_userName": "flands",
    },
    Object {
      "awsRegion": "ca-central-1",
      "eventID": "c9931287-acd4-478f-aa24-f77c8d297f71",
      "eventName": "DescribeSubnets",
      "eventSource": "ec2.amazonaws.com",
      "eventTime": "2018-06-14T17:17:56Z",
      "eventType": "AwsApiCall",
      "eventVersion": "1.05",
      "ingest_timestamp": 1538769915817,
      "log_type": "aws_cloud_trail",
      "recipientAccountId": "unittest9",
      "requestID": "03c4406b-42dc-4172-8beb-d76e15b9d69f",
      "responseElements": null,
      "sourceIPAddress": "13.56.178.189",
      "text": "{\\"eventVersion\\":\\"1.05\\",\\"eventTime\\":\\"2018-06-14T17:17:56Z\\",\\"eventSource\\":\\"ec2.amazonaws.com\\",\\"eventName\\":\\"DescribeSubnets\\",\\"awsRegion\\":\\"ca-central-1\\",\\"sourceIPAddress\\":\\"13.56.178.189\\",\\"userAgent\\":\\"aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8\\",\\"requestParameters\\":{\\"availabilityZoneSet\\":{}},\\"responseElements\\":null,\\"requestID\\":\\"03c4406b-42dc-4172-8beb-d76e15b9d69f\\",\\"eventID\\":\\"c9931287-acd4-478f-aa24-f77c8d297f71\\",\\"eventType\\":\\"AwsApiCall\\",\\"recipientAccountId\\":\\"unittest9\\"}",
      "userAgent": "aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8",
    },
    Object {
      "awsRegion": "ca-central-1",
      "eventID": "c9931287-acd4-478f-aa24-f77c8d297f71",
      "eventName": "DescribeSubnets",
      "eventSource": "ec2.amazonaws.com",
      "eventTime": "2018-06-14T17:17:56Z",
      "eventType": "AwsApiCall",
      "eventVersion": "1.05",
      "ingest_timestamp": 1538769915817,
      "log_type": "aws_cloud_trail",
      "recipientAccountId": "unittest9",
      "requestID": "03c4406b-42dc-4172-8beb-d76e15b9d69f",
      "requestParameters_bucketName": "myawsbucket",
      "requestParameters_host": Array [
        "s3.us-west-2.amazonaws.com",
      ],
      "requestParameters_versioning": Array [
        "",
      ],
      "responseElements": null,
      "sourceIPAddress": "13.56.178.189",
      "text": "{\\"eventVersion\\":\\"1.05\\",\\"userIdentity\\":{\\"type\\":\\"IAMUser\\",\\"principalId\\":\\"AIDAIRUNBBKT2UKVMBDMM\\",\\"arn\\":\\"arn:aws:iam::792677813823:user/flands\\",\\"accountId\\":\\"792677813823\\",\\"accessKeyId\\":\\"AKIAJDNI5LT67QRAUMIQ\\",\\"userName\\":\\"flands\\",\\"sessionContext\\":{\\"attributes\\":{\\"mfaAuthenticated\\":\\"false\\",\\"creationDate\\":\\"20131102T010628Z\\"},\\"sessionIssuer\\":{\\"type\\":\\"Role\\",\\"principalId\\":\\"AROAIDPPEZS35WEXAMPLE\\",\\"arn\\":\\"arn:aws:iam::123456789012:role/RoleToBeAssumed\\",\\"accountId\\":\\"123456789012\\",\\"userName\\":\\"RoleToBeAssumed\\"}}},\\"eventTime\\":\\"2018-06-14T17:17:56Z\\",\\"eventSource\\":\\"ec2.amazonaws.com\\",\\"eventName\\":\\"DescribeSubnets\\",\\"awsRegion\\":\\"ca-central-1\\",\\"sourceIPAddress\\":\\"13.56.178.189\\",\\"userAgent\\":\\"aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8\\",\\"requestParameters\\":{\\"host\\":[\\"s3.us-west-2.amazonaws.com\\"],\\"bucketName\\":\\"myawsbucket\\",\\"versioning\\":[\\"\\"]},\\"responseElements\\":null,\\"requestID\\":\\"03c4406b-42dc-4172-8beb-d76e15b9d69f\\",\\"eventID\\":\\"c9931287-acd4-478f-aa24-f77c8d297f71\\",\\"eventType\\":\\"AwsApiCall\\",\\"recipientAccountId\\":\\"unittest9\\"}",
      "userAgent": "aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8",
      "userIdentity_accessKeyId": "AKIAJDNI5LT67QRAUMIQ",
      "userIdentity_accountId": "792677813823",
      "userIdentity_arn": "arn:aws:iam::792677813823:user/flands",
      "userIdentity_principalId": "AIDAIRUNBBKT2UKVMBDMM",
      "userIdentity_sessionContext_attributes_creationDate": "20131102T010628Z",
      "userIdentity_sessionContext_attributes_mfaAuthenticated": "false",
      "userIdentity_sessionContext_sessionIssuer_accountId": "123456789012",
      "userIdentity_sessionContext_sessionIssuer_arn": "arn:aws:iam::123456789012:role/RoleToBeAssumed",
      "userIdentity_sessionContext_sessionIssuer_principalId": "AROAIDPPEZS35WEXAMPLE",
      "userIdentity_sessionContext_sessionIssuer_type": "Role",
      "userIdentity_sessionContext_sessionIssuer_userName": "RoleToBeAssumed",
      "userIdentity_type": "IAMUser",
      "userIdentity_userName": "flands",
    },
  ],
}
`;

exports[`CloudTrailKafkaCollector processLogsJson 1`] = `
Object {
  "logs": Array [
    Object {
      "awsRegion": "ca-central-1",
      "eventID": "c9931287-acd4-478f-aa24-f77c8d297f71",
      "eventName": "DescribeSubnets",
      "eventSource": "ec2.amazonaws.com",
      "eventTime": "2018-06-14T17:17:56Z",
      "eventType": "AwsApiCall",
      "eventVersion": "1.05",
      "ingest_timestamp": 1538769915817,
      "log_type": "aws_cloud_trail",
      "recipientAccountId": "unittest9",
      "requestID": "03c4406b-42dc-4172-8beb-d76e15b9d69f",
      "requestParameters_filterSet_items": Array [
        Object {
          "name": "vpc-id",
          "valueSet": Object {
            "items": Array [
              Object {
                "value": "vpc-04e3326d",
              },
            ],
          },
        },
      ],
      "responseElements": null,
      "sourceIPAddress": "13.56.178.189",
      "text": "{\\"eventVersion\\":\\"1.05\\",\\"userIdentity\\":{\\"type\\":\\"IAMUser\\",\\"principalId\\":\\"AIDAIRUNBBKT2UKVMBDMM\\",\\"arn\\":\\"arn:aws:iam::792677813823:user/flands\\",\\"accountId\\":\\"792677813823\\",\\"accessKeyId\\":\\"AKIAJDNI5LT67QRAUMIQ\\",\\"userName\\":\\"flands\\"},\\"eventTime\\":\\"2018-06-14T17:17:56Z\\",\\"eventSource\\":\\"ec2.amazonaws.com\\",\\"eventName\\":\\"DescribeSubnets\\",\\"awsRegion\\":\\"ca-central-1\\",\\"sourceIPAddress\\":\\"13.56.178.189\\",\\"userAgent\\":\\"aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8\\",\\"requestParameters\\":{\\"subnetSet\\":{},\\"filterSet\\":{\\"items\\":[{\\"name\\":\\"vpc-id\\",\\"valueSet\\":{\\"items\\":[{\\"value\\":\\"vpc-04e3326d\\"}]}}]}},\\"responseElements\\":null,\\"requestID\\":\\"03c4406b-42dc-4172-8beb-d76e15b9d69f\\",\\"eventID\\":\\"c9931287-acd4-478f-aa24-f77c8d297f71\\",\\"eventType\\":\\"AwsApiCall\\",\\"recipientAccountId\\":\\"unittest9\\"}",
      "userAgent": "aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8",
      "userIdentity_accessKeyId": "AKIAJDNI5LT67QRAUMIQ",
      "userIdentity_accountId": "792677813823",
      "userIdentity_arn": "arn:aws:iam::792677813823:user/flands",
      "userIdentity_principalId": "AIDAIRUNBBKT2UKVMBDMM",
      "userIdentity_type": "IAMUser",
      "userIdentity_userName": "flands",
    },
    Object {
      "awsRegion": "ca-central-1",
      "eventID": "c9931287-acd4-478f-aa24-f77c8d297f71",
      "eventName": "DescribeSubnets",
      "eventSource": "ec2.amazonaws.com",
      "eventTime": "2018-06-14T17:17:56Z",
      "eventType": "AwsApiCall",
      "eventVersion": "1.05",
      "ingest_timestamp": 1538769915817,
      "log_type": "aws_cloud_trail",
      "recipientAccountId": "unittest9",
      "requestID": "03c4406b-42dc-4172-8beb-d76e15b9d69f",
      "responseElements": null,
      "sourceIPAddress": "13.56.178.189",
      "text": "{\\"eventVersion\\":\\"1.05\\",\\"eventTime\\":\\"2018-06-14T17:17:56Z\\",\\"eventSource\\":\\"ec2.amazonaws.com\\",\\"eventName\\":\\"DescribeSubnets\\",\\"awsRegion\\":\\"ca-central-1\\",\\"sourceIPAddress\\":\\"13.56.178.189\\",\\"userAgent\\":\\"aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8\\",\\"requestParameters\\":{\\"availabilityZoneSet\\":{}},\\"responseElements\\":null,\\"requestID\\":\\"03c4406b-42dc-4172-8beb-d76e15b9d69f\\",\\"eventID\\":\\"c9931287-acd4-478f-aa24-f77c8d297f71\\",\\"eventType\\":\\"AwsApiCall\\",\\"recipientAccountId\\":\\"unittest9\\"}",
      "userAgent": "aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8",
    },
    Object {
      "awsRegion": "ca-central-1",
      "eventID": "c9931287-acd4-478f-aa24-f77c8d297f71",
      "eventName": "DescribeSubnets",
      "eventSource": "ec2.amazonaws.com",
      "eventTime": "2018-06-14T17:17:56Z",
      "eventType": "AwsApiCall",
      "eventVersion": "1.05",
      "ingest_timestamp": 1538769915817,
      "log_type": "aws_cloud_trail",
      "recipientAccountId": "unittest9",
      "requestID": "03c4406b-42dc-4172-8beb-d76e15b9d69f",
      "requestParameters_bucketName": "myawsbucket",
      "requestParameters_host": Array [
        "s3.us-west-2.amazonaws.com",
      ],
      "requestParameters_versioning": Array [
        "",
      ],
      "responseElements": null,
      "sourceIPAddress": "13.56.178.189",
      "text": "{\\"eventVersion\\":\\"1.05\\",\\"userIdentity\\":{\\"type\\":\\"IAMUser\\",\\"principalId\\":\\"AIDAIRUNBBKT2UKVMBDMM\\",\\"arn\\":\\"arn:aws:iam::792677813823:user/flands\\",\\"accountId\\":\\"792677813823\\",\\"accessKeyId\\":\\"AKIAJDNI5LT67QRAUMIQ\\",\\"userName\\":\\"flands\\",\\"sessionContext\\":{\\"attributes\\":{\\"mfaAuthenticated\\":\\"false\\",\\"creationDate\\":\\"20131102T010628Z\\"},\\"sessionIssuer\\":{\\"type\\":\\"Role\\",\\"principalId\\":\\"AROAIDPPEZS35WEXAMPLE\\",\\"arn\\":\\"arn:aws:iam::123456789012:role/RoleToBeAssumed\\",\\"accountId\\":\\"123456789012\\",\\"userName\\":\\"RoleToBeAssumed\\"}}},\\"eventTime\\":\\"2018-06-14T17:17:56Z\\",\\"eventSource\\":\\"ec2.amazonaws.com\\",\\"eventName\\":\\"DescribeSubnets\\",\\"awsRegion\\":\\"ca-central-1\\",\\"sourceIPAddress\\":\\"13.56.178.189\\",\\"userAgent\\":\\"aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8\\",\\"requestParameters\\":{\\"host\\":[\\"s3.us-west-2.amazonaws.com\\"],\\"bucketName\\":\\"myawsbucket\\",\\"versioning\\":[\\"\\"]},\\"responseElements\\":null,\\"requestID\\":\\"03c4406b-42dc-4172-8beb-d76e15b9d69f\\",\\"eventID\\":\\"c9931287-acd4-478f-aa24-f77c8d297f71\\",\\"eventType\\":\\"AwsApiCall\\",\\"recipientAccountId\\":\\"unittest9\\"}",
      "userAgent": "aws-sdk-java/1.11.170 Linux/4.4.121-k8s Java_HotSpot(TM)_64-Bit_Server_VM/25.162-b31/1.8.0_162 scala/2.11.8",
      "userIdentity_accessKeyId": "AKIAJDNI5LT67QRAUMIQ",
      "userIdentity_accountId": "792677813823",
      "userIdentity_arn": "arn:aws:iam::792677813823:user/flands",
      "userIdentity_principalId": "AIDAIRUNBBKT2UKVMBDMM",
      "userIdentity_sessionContext_attributes_creationDate": "20131102T010628Z",
      "userIdentity_sessionContext_attributes_mfaAuthenticated": "false",
      "userIdentity_sessionContext_sessionIssuer_accountId": "123456789012",
      "userIdentity_sessionContext_sessionIssuer_arn": "arn:aws:iam::123456789012:role/RoleToBeAssumed",
      "userIdentity_sessionContext_sessionIssuer_principalId": "AROAIDPPEZS35WEXAMPLE",
      "userIdentity_sessionContext_sessionIssuer_type": "Role",
      "userIdentity_sessionContext_sessionIssuer_userName": "RoleToBeAssumed",
      "userIdentity_type": "IAMUser",
      "userIdentity_userName": "flands",
    },
  ],
  "structure": "simple",
}
`;
